### PR TITLE
fix terminating thread kube

### PR DESCRIPTION
**Pull request by jrhizor**
Definitely not mergable, but I wanted to demonstrate one way we can do this. 

`HttpClientUtilsAirbyte` is a clone of `io.fabric8.kubernetes.client.utils.HttpClientUtils` except with the inclusion of the following line:
```
httpClientBuilder.connectionPool(new ConnectionPool(0, 30, TimeUnit.SECONDS));
```

Since it specifies that we can't keep idle connections around it will incur a small performance hit on the kube api server, but it would fix this problem for now at least. We could also allow idle connections but move the keep alive down to 1s/500ms/similar.

I tried just using 
```
OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder();
httpClientBuilder.connectionPool(new ConnectionPool(0, 30, TimeUnit.SECONDS));
httpClientBuilder.build();
```
but even in Docker for Desktop Kube it fails because it doesn't have all of that SSL configuration.

I bet there's an `OkHttp` client builder somewhere that is serviceable by default.

### CODE REVIEW

File: airbyte-workers/src/main/java/io/airbyte/workers/process/KubeProcessBuilderFactoryPOC.java [hunk 1]
```diff
@@ -95,9 +99,10 @@ public static void main(String[] args) throws InterruptedException, IOException
     // https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/#shutdown-isnt-necessary
     // Instead, the pod shuts down after 5 minutes as the pool reaps the remaining idle connection after
     // 5 minutes of inactivity, as per the default configuration.
+    KUBE_CLIENT.close();
+    OK_HTTP_CLIENT.dispatcher().cancelAll();
```
Comment (jrhizor) [inline airbyte-workers/src/main/java/io/airbyte/workers/process/KubeProcessBuilderFactoryPOC.java:103]: this had no impact

File: airbyte-workers/src/main/java/io/airbyte/workers/process/KubeProcessBuilderFactoryPOC.java [hunk 2]
```diff
@@ -36,21 +36,25 @@
...
-  // lies. See https://github.com/fabric8io/kubernetes-client/issues/2403.
-  private static final Config CONFIG = new ConfigBuilder().build();
-  private static final OkHttpClient OK_HTTP_CLIENT = HttpClientUtils.createHttpClient(CONFIG);
-  private static final KubernetesClient KUBE_CLIENT = new DefaultKubernetesClient(OK_HTTP_CLIENT, CONFIG);
 
   public static void main(String[] args) throws InterruptedException, IOException {
+    // Explicitly create the underlying HTTP client since the Kube client has issues with closing the
+    // client. It is not clear in which library the fault
+    // lies. See https://github.com/fabric8io/kubernetes-client/issues/2403.
+    final Config CONFIG = new ConfigBuilder().build();
```
Comment (jrhizor) [inline airbyte-workers/src/main/java/io/airbyte/workers/process/KubeProcessBuilderFactoryPOC.java:54]: making it non-static had no impact

File: airbyte-workers/src/main/java/io/airbyte/workers/process/KubeProcessBuilderFactoryPOC.java [hunk 3]
```diff
@@ -36,21 +36,25 @@
...
+    final OkHttpClient OK_HTTP_CLIENT = HttpClientUtilsAirbyte.createHttpClient(CONFIG);
```
Comment (jrhizor) [inline airbyte-workers/src/main/java/io/airbyte/workers/process/KubeProcessBuilderFactoryPOC.java:55]: this is the only piece that had impact

**Comment by davinchia** [comment]
nice! was thinking this is the 'best' way to do so.

I'm thinking we use `system.exit` for now, and open a PR against fabric to make their constructor configurable so we have less code to maintain. wdyt?

**Comment by jrhizor** [comment]
I'd actually prefer to clean this up slightly merge this in with a big comment linking to an issue and a fabric8 change. It looks like if it were accepted it'd take up to a month to get in place. We're going to need proper termination in a week. 

Either that or I think we'd need to wrap everything in our main loop with a try catch that `finally` calls system.exit? 

Either way I'll make the PR for fabric8 in the morning in hopes to get that into their next release ASAP.

**Comment by davinchia** [comment]
Makes sense. I'll create an issue for us and Fabric and use the system.exit version in the mean time.

I'm going to start using the cloud tag.

**Pull request closed without merging**